### DISCUSSION TITLE
Serve automatically generated static files with Nginx

### DISCUSSION THREAD

**Discussion by pooriamo** [vercel/next.js#11099] [Help]
Hi

Using `next build`, the automatically generated static files reside in `/project/.next/server/static/${BUILD_ID}/page.html`. 

The great thing is that we can write `try_files` for this file and serve it if present, But the problem is that the `BUILD_ID` is varied build to build and nginx doesn't know about it.

Is there a good way to handle this? 
Is using `try_files` the best option for this?

Thanks

**Comment by pooriamo** [comment]
Nothing? :)

**Comment by paambaati** [comment]
`try_files` is your best bet.

**Reply by pooriamo** [reply]
My only problem is the `BUILD_ID`

**Comment by pooriamo** [comment]
I think it's better to have this documented. You have created something great with a bad documentation. @timneutkens .

The doc says:

> If your pages don't have getInitialProps you may not need next export at all; next build is already enough thanks to Automatic Static Optimization.

Right, But how in the world can I serve this directory which has a variable path?!

**Reply by timneutkens** [reply]
> I think it's better to have this documented. You have created something great with a bad documentation. @timneutkens .

Thanks for the feedback, the documentation is open source, feel free to send a PR if you want to improve anything.

> Right, But how in the world can I serve this directory which has a variable path?!

Using `next start`. The deployment documentation covers deploying Next.js. If you use your own hosting it's `next build` and then `next start`.

**Reply by pooriamo** [reply]
I'm aware of it dear Tim. I've read the docs at least 3 times. I'm saying that what can I do to to DIRECTLY serve the generated html files with `nginx`. 

This is the scenario:
Dear nginx, please look for `index.html` in `/path_to_build_dir/.next/server/static/BUILD_ID/`. If it exists, serve it like React and Next never existed! 
If it doesn't proxy_pass to nextjs.

I hope I've described it well.

**Reply by timneutkens** [reply]
That's a custom setup and you're trying to premature optimization as you can cache the result from the Next.js server in nginx.

**Reply by pooriamo** [reply]
I want to absolutely eliminate unnecessary server calculations for those statically generated files.

**Reply by timneutkens** [reply]
Feel free to figure out the correct configuration and post it back here. As said in general we just recommend caching the response from the Next.js server.

**Reply by pooriamo** [reply]
@timneutkens I'll try. I have thought of generating a custom file by next.js on build time. This file can contain a one line code `try_files $uri $uri/ /path/.next/server/static/BUILD_ID/$uri.html @proxy_pass`

Nextjs can replace that BUILD_ID in each build and I can `include` this file in nginx. Does it make sense?

**Reply by timneutkens** [reply]
`.next/BUILD_ID` already holds the build id.

Note that this structure can change at any time hence why I'm saying it's better to use `next start`

**Reply by pooriamo** [reply]
@timneutkens Yes I know. But I think nginx doesn't support including text inside of a statement.
What you are saying needs this: `try_files $uri $uri/ /path/.next/server/static/{somehow include that BUILD_ID file}/$uri.html @proxy_pass` Which I believe is not supported by nginx.

And also reading a file and assigning the content to a variable is not supported AFAIK.

`next start` is the best option but it gives a redundant work to the server. Why my `express.js` (file: server.js) should evaluate and serve those html files to nginx and nginx pass it to the user?! While we can serve them directly by nginx.

**Comment by Palayoub** [comment]
Sad to read unfinished very important thing.

How can I serve my static generated files using "next export" to Nginx?

**Reply by kunKun-tx** [reply]
my current simple nginx config file for static generated files 

```
server {
    server_name www.example.com;
    root /www/example;
    index index.htm index.html;
        location /_next/static {
          add_header Cache-Control "max-age=31536000";
        }
        location / {
          try_files $uri.html  $uri $uri/ /index.html;
          add_header Cache-Control "no-cache";
        }

    listen [::]:80;
    listen 80; 
}
```

**Comment by davecardwell** [comment]
@pooriamo Something like this?

```nginx
location ~ ^/_next/static/(.*)$ {
  root /project/.next;
  try_files "/static/$1" "/server/static/$1" @proxy_pass;
}
```

**Reply by MohtashimAli85** [reply]
is $1 buildid ?
how we will be passing it in here?

**Reply by kumarldh** [reply]
`$1` is not build id, it is the first matched string in `location ~ ^/_next/static/(.*)$ `.  everything after `/_next/static/`.

**Comment by robbienohra** [comment]
@timneutkens quick question for ya

yesterday I actually pointed out a related typo in the docs that his since been corrected

https://github.com/vercel/next.js/discussions/21907

whereby the automatic optimization actually generates HTML in `.next/server/pages` and not `.next/server/static`

my question is this

I am actually using your suggested strategy of not using `try_files` but instead just caching at the nginx level

```
	location /_next/static {
		proxy_cache STATIC;
		proxy_pass http://nextjs:3000;
	}

	location /static {
		proxy_cache STATIC;
		proxy_ignore_headers Cache-Control;
		proxy_cache_valid 60m;
		proxy_pass http://nextjs:3000;
	}
```

does the typo I listed affect the path I should be using here? I've been trying to find more documentation on the nature of `_next`

any tips would be suggested thanks

**Reply by robbienohra** [reply]
@timneutkens @pooriamo the specific flavour of this problem I'm experiencing is that with my current nginx setup I'm still not seeing any HTML getting cached on the client-side anyways only JS files so I feel like I'm missing something under `_next` 🤔 

as I am seeing HTML files generated under `.next/server/pages` at build time

**Reply by robbienohra** [reply]
ah I think I answered my own question!

looks like HTML is served at the top level `/` with `_next/static` technically only for JS + CSS assets and not HTML

**Reply by Lmraza98** [reply]
Hi @robbienohra, would you happen to have a functional example of an nginx config for next? including the _next/static directory. Having some issues with getting it working.

**Reply by kumarldh** [reply]
@Lmraza98 this is a file which I am using for my local docker set up. Ignore the parts which doesn't concern you which is mostly my github userid and folder paths

```
# File has been built for Docker
# For easier way, few extra changes are needed

server {
        #Send all traffic to https
        listen 80 ;
        listen [::]:80 ;
        # CHANGE FOLLOWING TO SUITE YOUR ENVIRONMENT, ENSURE YOU SET THE SAME NAME IN YOUR /etc/hosts FILE
        server_name kumarldh.com;
        return 301 https://$host$request_uri;
}

server {
        # SSL configuration
        #
        listen 443 ssl ;
        listen [::]:443 ssl ;
        # THIS PATH ASSUMES THAT CERTIFICATE FOLDER IS AT /etc/certs, FOR EASIER WAY YOU WILL HAVE TO COPY IT TO THE SAID LOCATION
        ssl_certificate         /etc/certs/certbundle.pem;
        ssl_certificate_key     /etc/certs/star.kumarldh.com.key;

        # CHANGE FOLLOWING TO SUITE YOUR ENVIRONMENT, ENSURE YOU SET THE SAME NAME IN YOUR /etc/hosts FILE
        server_name kumarldh.com;
   # Serve all static files...
    location location ^~ /_next/static/ {
         root /usr/share/nginx/html/;
    }

    location / {
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-NginX-Proxy true;
        # IF YOU ARE RUNNING EASY WAY, UNCOMMENT FOLLOWING, AND MAKE CHANGES
        # proxy_pass http://your-fnp-frontend-HOST-name:3000;

        # IF YOU ARE NOT USING DOCKER, COMMENT LINE NUMBER #34
        # CHANGE PORT NUMBER TO SUITE YOUR ENVIRONMENT
        proxy_pass http://host.docker.internal:3000;

        proxy_ssl_session_reuse off;
        proxy_set_header Host $http_host;
        proxy_cache_bypass $http_upgrade;
        proxy_redirect off;
   }

}
```
The file above ☝️ is saved in `/Users/kumarldh/sandbox/dockers/nginx-docker/conf.d` as `default.conf` and the folder is mounted in nginx conf folder in below docker command/shell script 👇 
And this is my docker command/shell script
```
if [ "$(docker ps -aq -f name=mynginx)" ]; then
	docker start mynginx
else
	docker run --name=mynginx -d -p 80:80 -p 443:443 -v /Users/kumarldh/sandbox/dockers/nginx-docker/conf.d:/etc/nginx/conf.d -v /Users/kumarldh/sandbox/dockers/nginx-docker/certs:/etc/certs -v /Users/kumarldh/sandbox/nextjs-frontend-app/.next:/usr/share/nginx/html/_next nginx
fi
```

HTH!

**Comment by tomchentw** [comment]
For anyone who's interested, please be aware that the `_ssgManifest.js` file is generated via next.js build process in here: https://github.com/vercel/next.js/blob/v12.2.0/packages/next/build/index.ts#L2379-L2405
and it's **not generated by the webpack** build pipeline. Thus, you cannot use [compression-webpack-plugin](https://www.npmjs.com/package/compression-webpack-plugin) from webpack to compress assets into `.gz`, as this will compress an empty `_ssgManifest.js` and **deactivates** the prefetch feature of SSG.
